### ISSUE TITLE
how to create a shared private repository

### ISSUE THREAD

**Issue by redblade** [mikelangelo-project/capstan#85]
Hi, 
I'd like to setup a private shared repository that multiple users can use as their remote repo.
Basically a clone of the public repo on which to manage the upload of images by CLI (if feasible).
I could not find documentation about this, would it be possible and how?

Many thanks
Francesco

**Comment by miha-plesko** [comment]
Hi @redblade , it should be very easy to switch to your own repository. I've opened a PR to provide some more documentation about this ([link](https://github.com/mikelangelo-project/capstan/pull/86)), please let me know if there is something missing.

**Comment by gberginc** [comment]
@miha-plesko I think the main problem for @redblade currently is that they need a **private** shared repository. Because Capstan relies on publicly accessible S3 buckets, this is an unfortunate limitation of the current implementation.

Two questions for you @redblade, though

 * do you perhaps have an internal/private S3-compliant object storage service? If so, you could create an S3 bucket and ensure packages are structured as described in the open PR @miha-plesko mentioned
 * do you perhaps have a shared network file system installed? If so, you could put Capstan's local repository on NFS, accessible to all end-users. Then you would just need to set `$CAPSTAN_ROOT` env variable on all machines to point to the network share and work out of the pre-downloaded packages

**Comment by redblade** [comment]
Hi @gberginc, we don't have an S3 compliant service available but a plain http/https service is ok for our use-cases.
According to @miha-plesko it should be only a matter of creating two folders on an http/https service, set the new remote repository url, (copy the packages there) and pull the packages by name.
BTW, is there a command from CLI to push an artefact to the remote repository? 
Thanks again

**Comment by miha-plesko** [comment]
Unfortunately command like `capstan package push` is not available at the moment. Currently we're pushing base unikernel and all the packages to the S3 using this script:

https://github.com/mikelangelo-project/capstan-packages/blob/master/publish.sh

We can probably implement the `capstan package push` command for you if you depend on it, but we would need to know what mechanism is used for uploading?

**Comment by redblade** [comment]
First of all thanks for the availability!
I don't know if this is what you need: I think a basic upload on a http/https service would be enough;
later, the management of a basic authentication through let's say env variables could be a plus

**Comment by miha-plesko** [comment]
@redblade I don't think there is such thing as "basic upload to http/https service".  Usually http/https servers are used to serve content, not to receive one (which is great, because you don't want anyone out in the wild to be able to upload child porn stuff to your server 😈 ).

Basically it depends on your server's setup how files can be uploaded. If you were using S3 repository, we could support uploading files to S3 bucket. If you're using nginx to just serve files for you, then it's going to be a bit more complicated (perhaps we could employ FTP, but you need to configure that).

Anyway, may I ask you to open another issue for this so that we can close this one once the documentation PR is merged? I'm looking forward to continue discussion there!

**Comment by redblade** [comment]
ok, done ( #89 )

**Comment by redblade** [comment]
Thanks @miha-plesko , I have replicated the folder structure following your suggestion and I can now use it as a local repository. Uploading the packages in my folder is not a problem, a custom script will be enough or we can setup a basic (authenticated) http service to upload them if needed
